add BrokerSearch class for Headers retrieval

pavoljuhas opened the pull request:
In progress - needs test.  Comments welcome.

Implements https://github.com/NSLS-II/databroker-elasticsearch/issues/13.

pavoljuhas committed:
ENH: add BrokerSearch class for Headers retrieval

danielballan reviewed (commented):
I don't have all the context, but this seems reasonable to me.

pavoljuhas commented:
@danielballan - thanks for feedback and also for the email yesterday.

I am having an issue with testing this with an sqlite-based databroker.
The code needs to produce `Results` from a list of start document UIDs - [code here](https://github.com/pavoljuhas/databroker-elasticsearch/blob/c3a1244be2eba8bc8cedc14f9e1f5e7a42addef2/src/databroker_elasticsearch/brokersearch.py#L51-L52).

This however raises the following exception
```
AttributeError: 'HeaderSourceShim' object has no attribute 'run_start_given_uid'
```
Here is a minimum example that produces the same error
```python
import json
from databroker import Broker, temp_config
from databroker._core import Results

with open('src/tests/testdata/iss-sample.json') as fp:
    docs = json.load(fp)

cfg = temp_config()
db = Broker.from_config(cfg)
for d in docs:
    db.insert('start', d)

startstopuids = [(d['uid'], None) for d in docs]
res = Results(startstopuids, db, data_key=None)
print(list(res))
```

Is this because sqlite databroker does not implement `run_start_given_uid` or should I use some other way of creating `Results` from a sequence of start UIDs?

Should I switch to mongo-based broker?

pavoljuhas committed:
MNT: close parenthesis

pavoljuhas committed:
TST: suppress YAML load warnings from databroker

pavoljuhas committed:
TST: use `uid` for example elasticsearch

The `_id` cannot be reused for documents retrieved from databroker.

danielballan commented:
@tacaswell Any recollection of why this is missing from the shim?

pavoljuhas committed:
TST: cover BrokerSearch class

pavoljuhas committed:
BUG: fix arguments to ES bulk scan

Drop the `doc_type` argument which seem redundant.
Note: still failing on `Results` call.

pavoljuhas committed:
ENH: workaround bug in databroker.Results

Problem: Results class cannot be initialized from a sequence of uids.
Solution: Retrieve the run_start documents explicitly here.

pavoljuhas committed:
ENH: make `q` optional in BrokerSearch call

- match all entries with argument-less search
- support full Elasticsearch DSL-query with the `query` argument

pavoljuhas committed:
ENH: make `q` optional in ElasticIndex.qsearch

Match all with argument-less call.
Support argument `query` as a synonym for `body` for consistency
with BrokerSearch call interface.

pavoljuhas committed:
TST: verify more results from BrokerSearch

CJ-Wright commented on src/databroker_elasticsearch/brokersearch.py:
```diff
@@ -44,10 +46,10 @@ def __call__(self, q, **kwargs):
+        # FIXME: we need a fetchstart to work around bug in databroker.
+        # Remove the fetchstart mapping when fixed.
+        fetchstart = self.db.hs.mds.run_start_given_uid
+        gstartstop = ((fetchstart(e['_source']['uid']), None) for e in gscan)
```
Would this work with `db[e['_source']['uid']]`? If the uid is not in the databroker we might want this to fall over, since that is weird behavior. This would get around from digging 3 layers deep into the databroker (esp. since this low level syntax might move when databroker moves to intake but the top level syntax is more likely to be stable).

pavoljuhas replied:
As noted the fetchstart call is temporary and is needed to work around https://github.com/NSLS-II/databroker/issues/397.
Once that bug is fixed, fetchstart can be dropped and Results would be instantiated from a sequence of `(start_uid, None)` pairs.

Using `db[uid]` is less efficient as each Header then needs to be created twice.
That is because `db[uid]` is a `Header` but search should return a `Results` type, so each
Header would need to be split to a start/stop documents and re-created as a new Header in [`Results.__iter__`](https://github.com/NSLS-II/databroker/blob/171c98543d7a06deee2d8518e54d4709653880b3/databroker/_core.py#L857-L862) again.

pavoljuhas committed:
MNT: fix test so it passes in UTC time zone

pavoljuhas commented:
Ready for merge unless there are more suggestions.
